### ISSUE TITLE
tqdm object cannot be treated as bool when used with opened file

### ISSUE THREAD

**Issue by MarSoft** [tqdm/tqdm#353]
My use-case was to display upload progress when sending file with `POST` using `requests` library. I wrapped opened file with `tqdm`:
```python
f = open(filename, 'rb')
t = tqdm(f)
requests.post(url, data=t)
```
And I got an error:
```python
<...>
  File "/tmp/env/lib/python3.6/site-packages/requests/sessions.py", line 467, in request
    data = data or {},
  File "/tmp/env/lib/python3.6/site-packages/tqdm/_tqdm.py", line 753, in __len__
    else len(self.iterable)) if self.iterable is not None \
TypeError: object of type '_io.BufferedReader' has no len()
```

Turned out that `_io.BufferedReader` doesn't know file's length.
Okay, I passed file length to `tqdm` explicitly:
`t = tqdm(f, total=os.path.getsize(filename))`
Nothing changed.

Now, the problem happens when `requests` tries to treat `tqdm` as boolean to determine if there something passed or not. Tqdm has no `__bool__`, so Python falls back to `__len__`. And `__len__` is buggy in that it tries to get `len(self.iterable)` without checking if the iterable supports `__len__`.

I think it would be good to:
(1) return correct length if `total` was passed for non-measurable iterable; and probably
(2) implement separate `__bool__` which would return `True` if `__len__` returned `None` - e.g. if `tqdm` was passed an unmeasurable iterable and no explicit `total`. Or just return `True` always.

**Comment by casperdcl** [comment]
perhaps replace `tqdm(f)` with `tqdm(f.read())` or `tqdm(f.readlines())`?

**Comment by MarSoft** [comment]
@casperdcl, good luck trying this with 100GB file :-)
When I pass opened file to tqdm and then read it by 1k chunks, I use only small amount of RAM. And when I first read the whole file into string/binary/array and then pass resulting entity to tqdm, I use at least (size-of-file) RAM, which is often not acceptable.
So tqdm should work correctly with opened files, at least partially.

For a workaround I had to create a wrapper class which holds both original file and tqdm as its fields and correctly proxies all required methods, but it is an ugly solution.

**Comment by casperdcl** [comment]
yes, we never intended tqdm as a file wrapper but this should be easy enough to support. will open a PR

**Comment by casperdcl** [comment]
btw @MarSoft could you post your wrapper class code here? Just want to make sure we include your requirements

**Comment by MarSoft** [comment]
Actually, in my current implementation I used another approach: I created a wrapper class around `file` object which when iterated yields chunks of 1024 bytes, and then wrapped this class' instance with `tqdm`, explicitly specifying chunks count (`math.ceil(filesize/1024)`) and disabling `unit_scale`:
```
class ChunkedFile(object):
    """
    Regular open(name, 'rb'), when iterated, yields line-by-line.
    We want to ready by chunks.
    """
    def __init__(self, name, mode='rb', chunksize=1024):
        self.chunksize = chunksize
        self.length = op.getsize(name)
        self.f = open(name, mode)

    def __len__(self):
        return self.length

    def __iter__(self):
        return self

    def __next__(self):
        chunk = self.f.read(self.chunksize)
        if not chunk:
            raise StopIteration()
        return chunk

    def __enter__(self):
        return self

    def __exit__(self, *a):
        return self.f.__exit__(*a)
```
```
with ChunkedFile(filename, 'rb', chunksize=1024) as infile:
    ticker = tqdm(
        infile,
        unit='KB',
        unit_scale=False,
        total=math.ceil(infile.length / infile.chunksize),
    )
    requests.post(url, data=ticker)
```

**Comment by MarSoft** [comment]
Initially I didn't use chunking, but then upload turned out to be very slow.

**Comment by casperdcl** [comment]
if the only use cases for tqdm wrapping a file are 1) small files, so can `file.read()` into RAM, or 2) large file most likely requiring manual chunking anyway, then we probably should leave `tqdm` alone as-is. We might want to use your code in the examples/ folder instead.

**Comment by immerrr** [comment]
@casperdcl I have encountered this problem when trying to use tqdm over a size-less iterator. Apparently the situation has changed a bit since 2017, and it raises a different error, but this is still a problem:
```
In [19]: bool(tqdm.tqdm((1 for i in range(10))))                                                                                                                                              
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-19-e62055aa18f4> in <module>
----> 1 bool(tqdm.tqdm((1 for i in range(10))))

TypeError: 'NoneType' object cannot be interpreted as an integer

```

The problem is that the `tqdm` object does not have a `__bool__` (`__nonzero__` in Py2) operator and thus falls back to `__len__` when being cast to bool. And len being defined as
```
    def __len__(self):
        return self.total if self.iterable is None else \
            (self.iterable.shape[0] if hasattr(self.iterable, "shape")
             else len(self.iterable) if hasattr(self.iterable, "__len__")
             else getattr(self, "total", None))
```
fails fails because `self.total` is not initialized in the constructor and thus it is `None`, and Python does not like object's length being not a number.

It makes sense for a tqdm wrapped around an iterator not to have length, but it should be still good to use in `bool` context.

**Comment by casperdcl** [comment]
We could maybe something like `def __bool__(self): return bool(len(self)) if len(self) is not None else bool(self.iterable)`?

**Comment by immerrr** [comment]
@casperdcl please have a look at #694, it's implemented with the same general idea

**Comment by casperdcl** [comment]
yes saw and commented, will merge...

**Comment by casperdcl** [comment]
closed by #694
